### PR TITLE
chore: better cross-process communication

### PR DESCRIPTION
**Pull request by Rich-Harris**
Got nerd-sniped by @dummdidumm into finding a better way to run analysis/prerendering in a child process. This approach gives us type safety and type inference, and saves us messing around with `process.argv` nonsense. It's still a bit weird — there's no avoiding weirdness with this stuff, e.g. you have to pass `import.meta.url` as the first argument to `forked`, and you can only call it once per module — but I do think it's nicer than what's currently there.

### CODE REVIEW

File: packages/kit/src/core/postbuild/fork.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * @template T
+ * @template U
+ * @param {string} module
+ * @param {(opts: T) => U} callback
+ * @returns {(opts: T) => Promise<U>}
+ */
+export function forked(module, callback) {
+	if (process.env.SVELTEKIT_FORK) {
+		process.on(
```
Comment (benmccann) [inline packages/kit/src/core/postbuild/fork.js:13]: I thought you said we shouldn't do this: https://github.com/sveltejs/kit/pull/8636#issuecomment-1403816523

Comment (Rich-Harris) [reply packages/kit/src/core/postbuild/fork.js:13]: i did, but there isn't really a good alternative that doesn't involve a lot more faffery. by checking the `type` and the `module` we effectively guarantee that the only way we could get a false positive is if someone was deliberately trying to trip themselves up

File: packages/kit/src/core/postbuild/fork.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			'message',
+			/** @param {any} data */ async (data) => {
+				if (data.type === 'args' && data.module === module) {
```
Comment (Conduitry) [inline packages/kit/src/core/postbuild/fork.js:16]: Do we need to be more careful here? If the user is also sending messages, I don't think we can assume that `data` will be an object, can we?

```suggestion
				if (data && data.type === 'args' && data.module === module) {
```

File: packages/kit/src/core/postbuild/fork.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { fileURLToPath } from 'node:url';
+import child_process from 'node:child_process';
+
+/**
+ * @template T
+ * @template U
+ * @param {string} module
+ * @param {(opts: T) => U} callback
+ * @returns {(opts: T) => Promise<U>}
```
Comment (dummdidumm) [inline packages/kit/src/core/postbuild/fork.js:9]: I think this function deserves a bit of docs to understand how it works at a high level without diving into the code

```suggestion
 * Runs a task in a subprocess so any dangling stuff gets killed upon completion.
 * The subprocess needs to be the file `forked` is called in, and `forked` needs to be called eagerly at the top level.
 * @template T
 * @template U
 * @param {string} module `import.meta.url` of the file
 * @param {(opts: T) => U} callback The function that is invoked in the subprocess
 * @returns {(opts: T) => Promise<U>} A function that when called starts the subprocess
```

File: packages/kit/src/core/postbuild/prerender.js [hunk 4]
```diff
@@ -12,399 +11,404 @@ import { load_config } from '../config/index.js';
...
+	/**
+	 * @param {{
+	 *   out: string;
+	 *   manifest_path: string;
+	 *   metadata: import('types').ServerMetadata;
+	 *   verbose: boolean;
+	 *   env: Record<string, string>
+	 * }} opts
+	 */
+	async ({ out, manifest_path, metadata, verbose, env }) => {
```
Comment (dummdidumm) [inline packages/kit/src/core/postbuild/prerender.js:27]: Should we make this a function at the top level to safe one indentation level?

File: packages/kit/src/core/postbuild/analyse.js [hunk 5]
```diff
@@ -1,125 +1,131 @@
...
+import { forked } from './fork.js';
+
+export default forked(
+	import.meta.url,
+	/**
+	 * @param {{
+	 *   manifest_path: string;
+	 *   env: Record<string, string>
+	 * }} opts
+	 */ async ({ manifest_path, env }) => {
```
Comment (dummdidumm) [inline packages/kit/src/core/postbuild/analyse.js:19]: Should we make this a function at the top level to safe one indentation level?

**Review by dummdidumm — APPROVED** [review]
This is great! Left some small suggestions, but nothing serious

**Pull request merged**